numpy/core/include/numpy/ndarrayobject.h

```diff
@@ -232,9 +232,34 @@ PyArray_DiscardWritebackIfCopy(PyArrayObject *arr)
    dict.
 */
 
+#ifndef PYPY_VERSION
 #define NPY_TITLE_KEY(key, value) ((PyTuple_GET_SIZE((value))==3) && \
                                    (PyTuple_GET_ITEM((value), 2) == (key)))
-
+#else
+/* On PyPy, dictionary string keys do not always preserve object identity */
+static NPY_INLINE int
+NPY_TITLE_KEY_check(PyObject *key, PyObject *value)
+{
+    PyObject *title;
+    if (PyTuple_GET_SIZE(value) != 3) {
+        return 0;
+    }
+    title = PyTuple_GET_ITEM(value, 2);
+    if (key == title) {
+        return 1;
+    }
+    if (PyUnicode_Check(title) && PyUnicode_Check(key)) {
+        return PyUnicode_Compare(title, key) == 0;
+    }
+#if PY_VERSION_HEX < 0x03000000
+    if (PyString_Check(title) && PyString_Check(key)) {
+        return PyObject_Compare(title, key) == 0;
+    }
+#endif
+    return 0;
+}
+#define NPY_TITLE_KEY(key, value) (NPY_TITLE_KEY_check(key, value))
```

Comment:
Why the macro wrapping? Should both paths maybe be replaced with the inline function, which would narrow the scope of the PYPY check?

Reply:
Because there's code that does
`if NPY_TITLE_KEY(key, value) { ...` and omits the parentheses...

Reply:
Updated with comments explaining why.